internal/resourcestore/resourcecleaner.go

```diff
@@ -1,33 +1,72 @@
 package resourcestore
 
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/cri-o/cri-o/internal/log"
+)
+
 // A CleanupFunc is a function that cleans up one piece of
 // the associated resource.
-type CleanupFunc func()
+type CleanupFunc func() error
 
 // ResourceCleaner is a structure that tracks
 // how to cleanup a resource.
 // CleanupFuncs can be added to it, and it can be told to
 // Cleanup the resource
 type ResourceCleaner struct {
+	ctx   context.Context
 	funcs []CleanupFunc
 }
 
 // NewResourceCleaner creates a new ResourceCleaner
-func NewResourceCleaner() *ResourceCleaner {
+func NewResourceCleaner(ctx context.Context) *ResourceCleaner {
 	return &ResourceCleaner{
+		ctx:   ctx,
 		funcs: make([]CleanupFunc, 0),
 	}
 }
 
 // Add adds a new CleanupFunc to the ResourceCleaner
-func (r *ResourceCleaner) Add(f func()) {
-	r.funcs = append(r.funcs, CleanupFunc(f))
+func (r *ResourceCleaner) Add(f CleanupFunc) {
+	r.funcs = append(r.funcs, f)
 }
 
 // Cleanup cleans up the resource, running
 // the cleanup funcs in opposite chronological order
 func (r *ResourceCleaner) Cleanup() {
 	for i := len(r.funcs) - 1; i >= 0; i-- {
-		r.funcs[i]()
+		if err := retry(
+			r.funcs[i],
+		); err != nil {
+			log.Errorf(r.ctx, "Retried cleanup function too often, giving up")
+			// TODO: do we wanna return an error here?
+		}
+	}
+}
+
+// retry attempts to execute fn up to 5 times if its failure meets retryCondition.
+func retry(fn func() error) error {
+	backoff := wait.Backoff{
+		Duration: 500 * time.Millisecond,
+		Factor:   1.5,
+		Steps:    5,
 	}
+
+	waitErr := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		if err := fn(); err != nil {
+			return false, nil
```

Comment:
is there any way not to discard the real error returned from `fn()`?

Reply:
Can we log the error here?

Reply:
alright, done